README.md
pangaear
========



[![cran checks](https://cranchecks.info/badges/worst/pangaear)](https://cranchecks.info/pkgs/pangaear)
[![R-check](https://github.com/ropensci/pangaear/workflows/R-check/badge.svg)](https://github.com/ropensci/pangaear/actions?query=workflow%3AR-check)
[![codecov](https://codecov.io/gh/ropensci/pangaear/branch/master/graph/badge.svg)](https://codecov.io/gh/ropensci/pangaear)
[![rstudio mirror downloads](https://cranlogs.r-pkg.org/badges/pangaear)](https://github.com/r-hub/cranlogs.app)
[![cran version](https://www.r-pkg.org/badges/version/pangaear)](https://cran.r-project.org/package=pangaear)

`pangaear` is a data retrieval interface for the World Data Center PANGAEA (https://www.pangaea.de/). PANGAEA archieves published Earth & Environmental Science data under the following subjects: agriculture, atmosphere, biological classification, biosphere, chemistry, cryosphere, ecology, fisheries, geophysics, human dimensions, lakes & rives, land surface, lithosphere, oceans, and paleontology.

This package offers tools to interact with the PANGAEA Database, including functions for searching for data, fetching datasets by dataset ID, and working with the PANGAEA OAI-PMH service.

## Info

* Pangaea [website](https://www.pangaea.de/).
* Pangaea [OAI-PMH docs](https://wiki.pangaea.de/wiki/OAI-PMH).
* [OAI-PMH Spec](http://www.openarchives.org/OAI/openarchivesprotocol.html)

## Package API

 - pg_data
 - pg_list_metadata_formats
 - pg_identify
 - pg_list_records
 - pg_list_sets
 - pg_list_identifiers
 - pg_search
 - pg_get_record
 - pg_cache
 - pg_search_es
 - pg_cache_list
 - pg_cache_clear

## Installation

Stable version


```r
install.packages("pangaear")
```

Dev version


```r
remotes::install_github('ropensci/pangaear')
```


```r
library('pangaear')
```

## Search for data

This is a thin wrapper around the GUI search interface on the page <https://www.pangaea.de/>. Everything you can do there, you can do here.


```r
pg_search(query = 'water', bbox = c(-124.2, 41.8, -116.8, 46.1), count = 3)
#> # A tibble: 3 x 6
#>   score doi       size size_measure citation             supplement_to          
#>   <dbl> <chr>    <dbl> <chr>        <chr>                <chr>                  
#> 1 13.3  10.1594…     4 datasets     Krylova, EM; Sahlin… Krylova, EM; Sahling, …
#> 2 13.1  10.1594…     2 datasets     Simonyan, AV; Dultz… Simonyan, AV; Dultz, S…
#> 3  8.93 10.1594…   598 data points  WOCE Surface Veloci… <NA>
```

## Get data


```r
res <- pg_data(doi = '10.1594/PANGAEA.807580')
res[[1]]
#> <Pangaea data> 10.1594/PANGAEA.807580
#>   parent doi: 10.1594/PANGAEA.807580
#>   url:        https://doi.org/10.1594/PANGAEA.807580
#>   citation:   Schiebel, Ralf; Waniek, Joanna J; Bork, Matthias; Hemleben, Christoph (2001): Physical oceanography during METEOR cruise M36/6. PANGAEA, https://doi.org/10.1594/PANGAEA.807580, In supplement to: Schiebel, R et al. (2001): Planktic foraminiferal production stimulated by chlorophyll redistribution and entrainment of nutrients. Deep Sea Research Part I: Oceanographic Research Papers, 48(3), 721-740, https://doi.org/10.1016/S0967-0637(00)00065-0
#>   path:       /Users/sckott/Library/Caches/R/pangaear/10_1594_PANGAEA_807580.txt
#>   data:
#> # A tibble: 32,179 x 13
#>    Event       `Date/Time`   Latitude Longitude `Elevation [m]` `Depth water [m…
#>    <chr>       <chr>            <dbl>     <dbl>           <int>            <dbl>
#>  1 M36/6-CTD-… 1996-10-14T1…     49.0     -16.5           -4802             0   
#>  2 M36/6-CTD-… 1996-10-14T1…     49.0     -16.5           -4802             0.99
#>  3 M36/6-CTD-… 1996-10-14T1…     49.0     -16.5           -4802             1.98
#>  4 M36/6-CTD-… 1996-10-14T1…     49.0     -16.5           -4802             2.97
#>  5 M36/6-CTD-… 1996-10-14T1…     49.0     -16.5           -4802             3.96
#>  6 M36/6-CTD-… 1996-10-14T1…     49.0     -16.5           -4802             4.96
#>  7 M36/6-CTD-… 1996-10-14T1…     49.0     -16.5           -4802             5.95
#>  8 M36/6-CTD-… 1996-10-14T1…     49.0     -16.5           -4802             6.94
#>  9 M36/6-CTD-… 1996-10-14T1…     49.0     -16.5           -4802             7.93
#> 10 M36/6-CTD-… 1996-10-14T1…     49.0     -16.5           -4802             8.92
#> # … with 32,169 more rows, and 7 more variables: Press [dbar] <int>,
#> #   Temp [°C] <dbl>, Sal <dbl>, Tpot [°C] <dbl>, Sigma-theta [kg/m**3] <dbl>,
#> #   Sigma in situ [kg/m**3] <dbl>, Cond [mS/cm] <dbl>
```

Search for data then pass DOI to data function.


```r
res <- pg_search(query = 'water', bbox = c(-124.2, 41.8, -116.8, 46.1), count = 3)
pg_data(res$doi[3])[1:3]
#> [[1]]
#> <Pangaea data> 10.1594/PANGAEA.406110
#>   parent doi: 10.1594/PANGAEA.406110
#>   url:        https://doi.org/10.1594/PANGAEA.406110
#>   citation:   WOCE Surface Velocity Program, SVP (2006): Water temperature and current velocity from surface drifter SVP_9616641. PANGAEA, https://doi.org/10.1594/PANGAEA.406110
#>   path:       /Users/sckott/Library/Caches/R/pangaear/10_1594_PANGAEA_406110.txt
#>   data:
#> # A tibble: 101 x 10
#>    `Date/Time`  Latitude Longitude `Depth water [m… `Temp [°C]` `Cur vel U [cm/…
#>    <chr>           <dbl>     <dbl>            <int>       <dbl>            <dbl>
#>  1 1996-11-11T…     41.6     -125.                0        12.7            NA   
#>  2 1996-11-11T…     41.6     -125.                0        12.5            11.3 
#>  3 1996-11-12T…     41.6     -124.                0        12.4             2.91
#>  4 1996-11-12T…     41.7     -124.                0        12.3             3.64
#>  5 1996-11-12T…     41.7     -124.                0        11.9            23.4 
#>  6 1996-11-12T…     41.7     -124.                0        11.4            21.4 
#>  7 1996-11-13T…     41.8     -124.                0        11.1             0.21
#>  8 1996-11-13T…     41.8     -124.                0        11.2            -0.86
#>  9 1996-11-13T…     41.8     -124.                0        11.1             1.51
#> 10 1996-11-13T…     41.9     -124.                0        11.0            -5.58
#> # … with 91 more rows, and 4 more variables: Cur vel V [cm/s] <dbl>,
#> #   Latitude e <dbl>, Longitude e <dbl>, Code <int>
#> 
#> [[2]]
#> NULL
#> 
#> [[3]]
#> NULL
```

## OAI-PMH metadata


```r
# Identify the service
pg_identify()

# List metadata formats
pg_list_metadata_formats()

# List identifiers
pg_list_identifiers(from = Sys.Date() - 2, until = Sys.Date())

# List sets
pg_list_sets()

# List records
pg_list_records(from = Sys.Date() - 1, until = Sys.Date())

# Get a record
pg_get_record(identifier = "oai:pangaea.de:doi:10.1594/PANGAEA.788382")
```

## Contributors (reverse alphabetical)

* Naupaka Zimmerman
* Kara Woo
* Gavin Simpson
* Andrew MacDonald
* Scott Chamberlain

## Meta

* Please [report any issues or bugs](https://github.com/ropensci/pangaear/issues).
* License: MIT
* Get citation information for `pangaear` in R doing `citation(package = 'pangaear')`
* Please note that this package is released with a [Contributor Code of Conduct](https://ropensci.org/code-of-conduct/). By contributing to this project, you agree to abide by its terms.


NEWS.md
pangaear 1.1.0
==============

## MINOR IMPROVEMENTS

* add markdown pkg to Suggests for vignette


pangaear 1.0.0
==============

## MINOR IMPROVEMENTS

* vcr caching for pg_data tests, local only as files are too big (#74)
* vignette title fix (#73)
* `pg_search` fix for searching with a bounding box that crosses 180/-180 longitude: not a fix in this package, but the remote data source fixed a problem reported from a user of this package (#71)

pangaear 0.8.2
==============

## BUG FIXES

* detected from cran checks, only failing on debian clang r-devel: change `pg_search` tests to use `preserve_exact_body_bytes = TRUE` in the `vcr::use_cassette()` call so `yaml` package doesn't fail on loading it (#72)


pangaear 0.8.0
==============

## NEW FEATURES

* now using package `hoardr` for managing caching, replacing package `rappdirs`; new object `pg_cache`, an R6 class, with methods for managing where you cache files, deleting them, listing them, etc. Importantly, you can set the full cache path now, not just the folder within a set directory (#66) (#69)

## MINOR IMPROVEMENTS

* most tests using `vcr` now for http reqeust caching (#70)
* output of `pg_data()` now includes a `metadata` slot with parsed metadata from text files (only included when the file is a txt file); `parameters` slot that's part of the metadata is partially parsed into an unnamed list (#67)

## BUG FIXES

* fix in `pg_data()`: Pangaea changed how links are organized for datasets, fixed now (#65)
* fix in `pg_data()`: response content type header changed - a space was added, breaking the content type check; now not sensitive to the space (#68)


pangaear 0.6.0
==============

## MINOR IMPROVEMENTS

* Added a vignette (#62)
* replaced `httr` pkg with `crul` for HTTP requests (#55)
* Some datasets require login on the Pangaea platform - we now detect this in `pg_data` and skip the file download with a message saying so. Eventually we hope to fix this to allow uesrs to input credentials to get the file(s) (#59)


pangaear 0.3.0
==============

## NEW FEATURES

* New function `pg_search_es` - an interface to Pangaea's Elasticsearch
query interface.

## MINOR IMPROVEMENTS

* added more tests (#57)
* now using markdown in docs 
* tidy code to 80 line width

## BUG FIXES

* fixed bug in oai functions due to changed base url for the 
Pangaea OAI server (#53)
* Fix to `pg_search` as search portal now has offset param - so if
more than 500 results need to page through them with the 
`offset` parameter (#56)


pangaear 0.2.4
==============

## MINOR IMPROVEMENTS

* Improved examples in the OAI methods to make sure they work 
regardless of when the user runs them

## BUG FIXES

* Fixes to `pg_search()` needed due to changes in the Pangaea 
website. Nearly identical functionality, but one parameter switch
(`env` param is now `topic`). More parameters may be added in the 
future. New fields are added to output, added to docs for the 
function. Now importing `jsonlite` as a result of these changes (#50)
* Fixes to `pg_data()` needed due to changes in the Pangaea 
website. Nothing should be different for the user. (#51)


pangaear 0.2.0
==============

## MINOR IMPROVEMENTS

* Dropped `XML`, using `xml2` now (#42)
* Using `rappdirs` package now for determing caching path on user's machine (#46)
* using `tibble` now for compact data.frame representations (#47)
* Dropped `methods` dependency (#48)
* Added test suite (#45)

## BUG FIXES

* Fixes to `pg_search()` - introduced due to changes in Pangaea website (#44)

pangaear 0.1.0
==============

* Released to CRAN.


R/pg_search.R
#' Search the Pangaea database
#'
#' @export
#' @param query (character) Query terms. You can refine a search by prefixing
#' the term(s) with a category, one of citation, reference, parameter, event,
#' project, campaign, or basis. See examples.
#' @param topic (character) topic area: one of NULL (all areas), "Agriculture",
#' "Atomosphere", "Biological Classification", "Biospshere", "Chemistry",
#' "Cryosphere", "Ecology", "Fisheries", "Geophysics", "Human Dimensions",
#' "Lakes & Rivers", "Land Surface", "Lithosphere", "Oceans", "Paleontology"
#' @param count (integer) Number of items to return. Default: 10. Maximum: 500.
#' Use `offset` parameter to page through results - see examples
#' @param offset (integer) Record number to start at. Default: 0
#' @param bbox  (numeric) A bounding box, of the form: minlon, minlat, maxlon,
#' maxlat
#' @param mindate,maxdate (character) Dates to search for, of the form
#' "2014-10-28"
#' @param ... Curl options passed on to [crul::verb-GET]
#' @return tibble/data.frame with the structure:
#' 
#' - score: match score, higher is a better match
#' - doi: the DOI for the data package
#' - size: size number
#' - size_measure: size measure, one of "data points" or "datasets"
#' - citation: citation for the data package
#' - supplement_to: citation for what the data package is a supplement to
#' 
#' @details This is a thin wrapper around the GUI search interface on the page
#' <https://www.pangaea.de>. Everything you can do there, you can do here.
#' @seealso [pg_search_es()]
#' @examples \dontrun{
#' pg_search(query='water')
#' pg_search(query='water', count=2)
#' pg_search(query='water', count=20)
#' pg_search(query='water', mindate="2013-06-01", maxdate="2013-07-01")
#' pg_search(query='water', bbox=c(-124.2, 41.8, -116.8, 46.1))
#' pg_search(query='reference:Archer')
#' pg_search(query='parameter:"carbon dioxide"')
#' pg_search(query='event:M2-track')
#' pg_search(query='event:TT011_2-CTD31')
#' pg_search(query='project:Joint Global Ocean Flux Study')
#' pg_search(query='campaign:M2')
#' pg_search(query='basis:Meteor')
#'
#' # paging with count and offset
#' # max is 500 records per request - if you need > 500, use offset and count
#' res1 <- pg_search(query = "florisphaera", count = 500, offset = 0)
#' res2 <- pg_search(query = "florisphaera", count = 500, offset = 500)
#' res3 <- pg_search(query = "florisphaera", count = 500, offset = 1000)
#' do.call("rbind.data.frame", list(res1, res2, res3))
#'
#' # get attributes: maxScore, totalCount, and offset
#' res <- pg_search(query='water', bbox=c(-124.2, 41.8, -116.8, 46.1))
#' attributes(res)
#' attr(res, "maxScore")
#' attr(res, "totalCount")
#' attr(res, "offset")
#'
#' # curl options
#' pg_search(query='citation:Archer', verbose = TRUE)
#' }

pg_search <- function(query, count = 10, offset = 0, topic = NULL, bbox = NULL,
                      mindate = NULL, maxdate = NULL, ...) {
  calls <- names(sapply(match.call(), deparse))[-1]
  calls_vec <- "env" %in% calls
  if (any(calls_vec)) {
    stop("'env' has been removed, use topic instead and see ?pg_search",
         call. = FALSE)
  }

  check_if(count, c("numeric", "integer"))
  check_if(topic, "character")
  check_if(mindate, "character")
  check_if(maxdate, "character")
  args <- pgc(list(t = topic, count = count, offset = offset, q = query,
                   mindate = mindate, maxdate = maxdate))
  if (!is.null(bbox)) args <- c(
    args, as.list(stats::setNames(bbox,
                                  c('minlon', 'minlat', 'maxlon', 'maxlat'))))
  
  cli <- crul::HttpClient$new(url = sbase())
  res <- cli$get(query = args, ...)
  res$raise_for_status()
  results <- jsonlite::fromJSON(res$parse("UTF-8"), FALSE)
  parsed <- lapply(results$results, function(x) {
    x <- utils::modifyList(x, list(doi = gsub("doi:", "", x$URI)))
    xx <- parse_res(x)
    x$URI <- x$html <- NULL
    c(x, xx)
  })
  df <- do.call("rbind.data.frame", lapply(parsed, tibble::as_tibble))
  atts <- results[c('maxScore', 'offset', 'totalCount')]
  for (i in seq_along(atts)) {
    attr(df, names(atts)[i]) <- atts[[i]]
  }
  return(df)
}

parse_res <- function(x) {
  html <- xml2::read_html(x$html)
  citation <- xml_text(xml_find_all(html, './/div[@class="citation"]/a'))
  tab <- xml_find_all(html, './/table/tr')
  supp <- xml_text(
    xml_find_first(
      xml_parent(
        xml_find_all(
          tab, ".//td[contains(.,'Supplement')]")), './/td[@class="content"]'))
  size <- strextract(
    xml_text(
      xml_find_all(
        xml_parent(
          xml_find_all(
            tab,
            ".//td[contains(.,'Size')]")),
        './/td[@class="content"]')), ".+")
  size_val <- strextract(size, "[0-9]+")
  size_val2 <- tryCatch(as.numeric(size_val), warning = function(w) w)
  size_val <- if (inherits(size_val2, "warning"))  size_val else size_val2
  meas <- strextract(size, "[A-Za-z].+")
  lis <- list(
    size = size_val,
    size_measure = meas,
    citation = citation,
    supplement_to = supp
  )
  lis[vapply(lis, length, 1) == 0] <- NA
  lis
}

check_if <- function(x, cls) {
  if (!is.null(x)) {
    if (!class(x) %in% cls) {
      stop(substitute(x), " must be of class: ", paste0(cls, collapse = ", "),
           call. = FALSE)
    }
  }
}
